perf_capture_timer preload resources to reduce queries

kbrock opened the pull request:
We have many queries for perf_capture_timer. This aims to reduce them.

Still have one more pass after this

- We were looking up `Zone` multiple times.
- We were loading all the vms without an availability zone even when not necessary.
- Made a number of smaller changes, like using `flat_map` and `with_object`, which will have a negligible effect.
- Split up the primary method so we could focus on the slower portion of code.

||before|after
---|---|---
vms|10k|10k|
miq_queue|7950|7950|
miq_task|1543|7950|
time|6750ms|4729 ms
sql count|879|679
sql time|323ms|191ms
memory|190mb|110mb

**DISCLAIMER:** this does not include `capture_queue` (which is the biggest method) (it was timing out on my box)

https://bugzilla.redhat.com/show_bug.cgi?id=1221750

Fryguy commented on app/models/metric/targets.rb:
```diff
@@ -19,7 +19,8 @@ def self.capture_infra_targets(zone, options = {})
 
     targets = targets.select(&:perf_capture_enabled?)
-    targets = targets.collect { |t| t.kind_of?(EmsCluster) ? t.hosts : t }.flatten
+    MiqPreloader.preload(targets.select { |t| t.kind_of?(EmsCluster) }, :hosts)
```
This is the N+1 fix?  Would It be simpler to just adjust the includes on line 14 to `:ems_clusters => [:hosts, :tags]`?

Fryguy commented on app/models/zone.rb:
```diff
@@ -161,10 +161,9 @@ def availability_zones
-    ext_management_systems.flat_map do |e|
-      e.kind_of?(EmsCloud) ? e.vms.select { |vm| vm.availability_zone.nil? } : []
-    end
+    clouds = ext_management_systems.select { |e| e.kind_of?(EmsCloud) }
```
I think there's an `ems_clouds` association that you can use here instead of `ext_management_systems`, however that might actually *add* more overhead...not sure.

kbrock commented:
@Fryguy that first suggestion was beautiful - will check second one (think it will be worse though)

@jrafanie thanks for keeping me honest. document all things!

Fryguy commented on app/models/metric/capture.rb:
```diff
@@ -78,7 +72,10 @@ def self.perf_capture_timer(zone = nil)
+    targets_by_rollup_parent
+  end
 
+  def self.calc_tasks_by_rollup_parent(targets_by_rollup_parent)
```
Can these new methods be made into private methods?  They shouldn't be part of the public interface, I don't think.

Fryguy commented on app/models/metric/capture.rb:
```diff
@@ -43,8 +43,7 @@ def self.capture_threshold(target)
-  def self.perf_capture_health_check(zone = nil)
-    zone ||= MiqServer.my_server.zone(true)
+  def self.perf_capture_health_check(zone)
     q_items = MiqQueue.select("created_on, args").where(:state => "ready", :role => "ems_metrics_collector", :method_name => "perf_capture", :zone => zone.name).order("created_on ASC")
```
The order can just be `order(:created_on)` ... ASC is default.

Fryguy replied:
Oh nvm...you didn't touch this line (diff-reading-fail)

kbrock replied:
I want to change this method so badly :)

Fryguy commented:
This looks great @kbrock.  Let's get it in as is without any more changes.  These are very atomic and solve the problem...not need to clutter up the finished good decisions with more changes...you can always make a second PR.

I made a minor comment about making those new methods private.  Once that's in, this is good to go.

Fryguy commented on app/models/metric/capture.rb:
```diff
@@ -43,8 +43,7 @@ def self.capture_threshold(target)
 
-  def self.perf_capture_health_check(zone = nil)
-    zone ||= MiqServer.my_server.zone(true)
+  def self.perf_capture_health_check(zone)
```
I feel like this method should be private as well.  We don't call it from anywhere else, but perf_capture_timer.

kbrock commented:
@Fryguy updated

- I moved the code around to put the privates together. Was tricky to make the class methods private.
- Fun how the methods I extracted are too complex. What about the original? ;)
- Fixed specs.

jrafanie commented:
:+1: Very nice.
